### PR TITLE
Drop Spark-2 support and update dependencies

### LINKED ISSUE #202: Drop support for Python 3.7
Drop support for Python 3.7 as it is EOL and upgrade Poetry dependencies to >= Python 3.8

### PR DESCRIPTION
**Pull request by SemyonSinchenko**
## Proposed changes

- Update deps
- Update Ruff
- Corresponding updates of pyproject
- Slightly update Makefile
- Drop Support of Spark2
- Drop Support of Spark 3.1-3.2
- Minimal python is 3.10 from now
- Apply new ruff rules
- Apply ruff linter

## Types of changes

What types of changes does your code introduce to quinn?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Further comments

Close #202

### CODE REVIEW

**Comment by MrPowers** [comment]
Looks like PySpark 2 downloads are a small percentage of overall downloads now:

Screenshot 2024-07-14 at 9 11 24 AM

So now is a good time to drop PySpark 2 support IMO.

**Comment by MrPowers** [comment]
LGTM, let me know if we're ready to merge!

File: pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "quinn"
-version = "0.10.3"
+version = "1.0"
```
Comment (jeffbrennan) [inline pyproject.toml:3]: This should be 1.0.0 for semver right

Comment (SemyonSinchenko) [reply pyproject.toml]: Done!

**Comment by jeffbrennan** [comment]
LGTM! Thanks for working on this

**Pull request merged**